### PR TITLE
[otbn] Better specify and describe errors

### LINKED ISSUE #8238: [otbn] Define a list of errors to align ERR_BITS and FATAL_ALERT_CAUSE
We currently have two external CSRs containing error information, `ERR_BITS` and `FATAL_ALERT_CAUSE`. `ERR_BITS` is only updated when an operation (one started by writing `CMD`) fails, while `FATAL_ALERT_CAUSE` only contains fatal errors.

Since we defined these registers we made smaller changes to the spec and the implementation, creating a bit of inconsistency and sources for confusion. For example, we expanded the term "operation" to not only mean "execute code", but also "execute a secure wipe operation". That means `ERR_BITS` can now also be set as result of a data or instruction memory secure wipe operation. [That's not yet implemented, but specified.]

To reduce the confusion a bit, here's a proposed list of action:
* Create a table with all errors (which also includes a unique error name).
* Update the documentation on `ERR_BITS` and `FATAL_ALERT_CAUSE` to reference these errors, instead of defining two lists of similar errors.

I had a look at the errors we have today and created a quick spreadsheet at 
https://docs.google.com/spreadsheets/d/1zLSFSWAY6VetNSDRvmVkZAUdMlJU5E5YfygpsIN1yZo/edit?usp=sharing

@rswarbrick @GregAC WDYT? If you're OK with the general course of action, I'll follow up with a PR.

### PR DESCRIPTION
**Pull request by imphil**
Introduce a list of all errors that OTBN detects and improve their
description to better describe what caused the error. Reference this
list as "authorative source" in other places, e.g. in the register
table.

Fixes #8238

### CODE REVIEW

File: hw/ip/otbn/doc/_index.md [hunk 1]
```diff
@@ -630,10 +630,93 @@ However, every error that OTBN detects when it isn't running is fatal.
...
+        One of the following things happened:
+        <ul>
+          <li>A loop was started with an iteration count of zero.</li>
+          <li>The final instruction of a loop was a branch or another loop.</li>
+          <li>A new loop tried to push to a full loop stack (loop nesting level too deep).</li>
+        </ul>
+      </td>
+    </tr>
+    <tr>
+      <td><code>FATAL_IMEM<code></td>
```
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:684]: This naming is inconsistent with the naming in FATAL_ALERT_CAUSE, and at the same time, doesn't mention that we're talking about integrity errors here.

I'd propose to rename this error to `IMEM_INTG_VIOLATION`. Alternatively, we can make it slightly broader to cover all kinds of "imem access problems" (maybe the SRAM macro will give us errors as well?) and rename to `IMEM_ACCESS_ERROR` or something like that?

File: hw/ip/otbn/doc/_index.md [hunk 2]
```diff
@@ -630,10 +630,93 @@ However, every error that OTBN detects when it isn't running is fatal.
...
+    <tr>
+      <td><code>ILLEGAL_INSN<code></td>
+      <td>recoverable</td>
+      <td>
+        An illegal instruction was about to be executed.
+        An instruction is illegal in the following cases:
+        <ul>
+          <li>An instruction being excuted had an invalid encoding.</li>
+          <li>An access occurred for an invalid CSR or WSR.</li>
+        </ul>
```
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:668]: I'm wondering if we should remove this detailed list and refer to the ISA documentation instead, which now lists the error conditions in much more detail. (Here and below in the LOOP case.)

File: hw/ip/otbn/data/otbn.hjson [hunk 3]
```diff
@@ -364,7 +337,7 @@
     } // register : start_addr
     { name: "FATAL_ALERT_CAUSE",
```
Comment (imphil) [inline hw/ip/otbn/data/otbn.hjson:338]: Once we have decided on a naming for the errors, I'd align the names in FATAL_ALERT_CAUSE with the ones in ERR_BITS.

File: hw/ip/otbn/doc/_index.md (hunk 1 above)
Comment (imphil) [inline hw/ip/otbn/doc/_index.md:684]: Another topic: I'd probably remove the "FATAL_" prefix from the error names. We now have this information in the table itself. And with the upcoming feature to promote recoverable errors to fatal errors, having the prefix seems more confusing than helping.

**Comment by imphil** [comment]
@GregAC @rswarbrick I left a couple comments for discussion items. PTAL if you have a moment and let me know what you think. Particularly contentious items can be discussed in the Monday sprint meeting :-)

**Comment by imphil** [comment]
And one more topic that came up when reviewing #8146:

How about calling all current "recoverable errors" "software errors" instead? A software error is an error which (in the absence of an attacker) is present in the software code that OTBN runs, e.g. a wrong encoding, an out of bounds DMEM access, etc. Software errors are recoverable by default, but can be promoted to be fatal (#8203). All "software errors" are listed in the ISA spec (as extended in #8146).

File: hw/ip/otbn/doc/_index.md (hunk 2 above)
Comment (rswarbrick) [reply hw/ip/otbn/doc/_index.md:668]: Sounds like a good idea to me.

File: hw/ip/otbn/doc/_index.md (hunk 1 above)
Comment (rswarbrick) [reply hw/ip/otbn/doc/_index.md:684]: Is it just a namespacing thing to avoid clashes in the RTL? If not, I'm happy to drop it.

**Review by rswarbrick — APPROVED** [review]

**Comment by rswarbrick** [comment]
Also, I like the recoverable -> software change. Sounds like a good idea to me!

File: hw/ip/otbn/doc/_index.md (hunk 1 above)
Comment (GregAC) [reply hw/ip/otbn/doc/_index.md:684]: No issues with dropping it

**Review by GregAC — APPROVED** [review]
A very sensible change. Also in favour of switching 'recoverable' to 'software'

**Comment by imphil** [comment]
@GregAC @rswarbrick I made the changes as discussed. Would you mind having another look? Thanks!

**Review by rswarbrick — APPROVED** [review]

**Review by GregAC — APPROVED** [review]

**Pull request merged**